abdiff/core/calc_ab_metrics.py

```diff
@@ -65,22 +65,35 @@ def create_record_diff_matrix_dataset(
     for i, batch in enumerate(
         diffs_ds.to_batches(
             batch_size=batch_size,
-            columns=["timdex_record_id", "source", "modified_timdex_fields", "has_diff"],
+            columns=[
+                "timdex_record_id",
+                "source",
+                "modified_timdex_fields",
+                "has_diff",
+                "record_a",
+                "record_b",
+            ],
         )
     ):
         start_time = time.time()
         batch_df = batch.to_pandas()
 
         batch_metrics = []
         for _, row in batch_df.iterrows():
+            has_diff = 1 if row["has_diff"] == "true" else 0
+            a_or_b_missing = (
+                1 if (pd.isna(row["record_a"]) or pd.isna(row["record_b"])) else 0
+            )
```

Comment:
Hmm, do we also need to account for when `a_and_b_missing`? 🤔

Reply:
I guess I don't totally follow?  This is where we first set that field, by setting it as true when either record a or b is missing.

Reply:
Ah, I guess I was referring to how in `calc_ab_diffs`, we [explicitly check whether `a` **and** `b` is missing vs. `a` **or** `b` is missing -- and if we needed a column in the metrics dataset to denote that difference as well](https://github.com/MITLibraries/transmogrifier-ab-diff/pull/65/files#diff-f6d81437fc3dd591a024246e6ea6c0b3093e76e4c16e3cb4a018736c22bf0924R138-R141)?

Reply:
Yeah, I waffled a bit on the order of when we detect vs record vs handle the absence of missing record.

The approach taken here is:
- `collate_ab_transforms`
  - will happily join a record where either A and/or B is absent, doing nothing with this information
- `calc_ab_diffs`
  - the A/B records are passed as-is to sub-function `calc_record_diff`
  - this notices when A and/or B is missing, and returns the tuple `(None, set(), True|False)` depending if one or both
  - at this point, we haven't really recorded that A and/or B is missing, just that we can't really do a meaningful record-level diff
- `calc_ab_metrics`
 - here we finally set a column in a dataset, `a_or_b_missing` which indicates if one or both are missing
 - we know that this will get included in the final `records` dataset (merging of `diffs` and `metrics` datasets)

At that point, we _still_ don't do anything with this information, but the webapp could surface this in a meaningful way.  Or, we could add some more analysis [around here]() to include that in higher level summary statistics (e.g. count of records where both missing, count of records where A missing, count of records where B missing, etc.).  We could also surface on the Record page itself, but it'd be pretty obvious from looking at the A and B versions.

I had considered adding this `a_or_b_missing` at the `calc_ab_diffs` stage, or even the collating... but those would have been more invasive changes, and I wasn't seeing much benefit.  As-is, the "records-with-missing-something" kind of happily march through the steps, so they populate just as normal the datasets as they get created.  It's not until late in `calc_ab_metrics` that we finally note it, and now the decision is how do we surface that information.  But I'd propose that's work for another ticket 😎.

Reply:
Thank you for breaking down how the flow of information between `calc_ab_diffs` and `calc_ab_metrics`! I understand it a lot better now! Your idea to include a note about missingness in "higher level summary statistics" sounds great but outside the scope of this ticket!